### ISSUE TITLE
Quicken FastMovieWriter?

### ISSUE THREAD

**Issue by adehad** [cculianu/StimulateOpenGL_II#1]
Just inquiring if there are any plans to improve the speed of FastMovieWriter?
At the moment we are able to add a frame every 5 seconds, which results in tediously long `.fmv` creation time for our high frame rate system.

I had a look at some of the source code and it seems that speed ups could be done in:
1. Having a constructor that takes the (maximum) number of frames as an argument, to then pre-allocate space for all the frames.
2. Utilising asynchronous running of the file writing commands compared to the MATLAB script. As demonstrated here for example: https://uk.mathworks.com/help/matlab/matlab_external/call-matlab-from-user-threads-in-mex-function.html  

I would also be excited to try these out myself, but cannot seem to find current build instructions?

Cheers

**Comment by cculianu** [comment]
Hi there!

Hmm.. I'll take a look at that stuff.  Yes, I probably can be improved dramatically, for sure. I think when I developed this there was no threading in matlab... 

5 seconds per frame is abysmal. Wow.  Is it spending most of its time in FastMovieWriter ??

**Comment by cculianu** [comment]
Wow that link you pasted is very sexy C++ code.  I hadn't realized matlab updated their api to be so C++-friendly.

As for build instructions -- I forgot how to build this thing! I seem to remember mostly the matlab bits required just "mex file.c" or whatever -- is that not working?

**Comment by adehad** [comment]
Indeed it appears to spend a majority of its time in AddFrame:
![image](https://user-images.githubusercontent.com/26027314/58752442-accc0480-84a6-11e9-92cc-55325c742e35.png)

However, I will be doing more Profiling in the coming days to see if there is any 'badness' on our side.

Ahaha, yes MATLAB does have quite a few useful features that are terribly documented. If you are interested, there is an article by the godly Yair Altman, which you might find interesting:
https://undocumentedmatlab.com/blog/multi-threaded-mex

I did try the `mex file.cpp` command, but it seems that with the MS Build tools I was using, it didn't play nice.

**Comment by adehad** [comment]
I believe I found the cause, the `uint8` conversion done inside the MEX file!
https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/FastMovieWriterMex/FastMovieWriterMex.cpp#L168-L205

Might be better to do it in the MATLAB call itself.
https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/%40FastMovieWriter/AddFrame.m#L25

e.g. 
`FastMovieWriterMex('addFrame', g.handle, uint8(frame), clevel);`

Or throw a warning of sorts.

The Profiler now says `AddFrame` takes 0.18s for our entire stimulus (!)

I found some other inefficiencies in our main script, so I'll look into those to further speed up our rendering.

**Comment by cculianu** [comment]
Nice job!!  Oh wow. That should be way better documented.  You are right -- it should throw an exception.  It's useless to accept those values if the slowdown is so large.

Thanks for catching this.  I honestly forgot how this code works -- I'm super glad you managed to get it working!!

**Comment by adehad** [comment]
So I just found out that `AddFrame` does not take 0.18s for the same stimulus used in the screenshot earlier, my bad.
The profile does show we saved 300s or so of run time though (using MATLAB to typecast to `uint8`):
![image](https://user-images.githubusercontent.com/26027314/58767541-867c9680-8584-11e9-9811-d310c8dd0ef4.png)

So it still might be worth looking into further optimisations.

**Comment by cculianu** [comment]
Hmm. It might be spending most of its time compressing.  Try passing a third argument to AddFrame, `0`.  This is the compression level. Try turning off compression altogether.  See this comment:

https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/%40FastMovieWriter/AddFrame.m#L5

I think if compression is set to 0, it should be very fast.

**Comment by adehad** [comment]
It is indeed faster (around 400s). However the filesize went from ~30MB to ~30GB (!!!).
![image](https://user-images.githubusercontent.com/26027314/58768828-b5046c80-8598-11e9-8b1f-a8ad5959113d.png)

**Comment by cculianu** [comment]
So try compression 1, 2, 3, etc.  9 is the maximum (and the default!).  I bet even with 1 or 2 it will be much better.

Sadly I should have used a real movie format to do this -- a lossless one such as FFv1.  But at the time integrating a real movie encoder like ffmpeg was beyond the scope of this project.

I ~~suspect~~ hope with compression 1 or 2 or 3 you may get a happy balance between speed and size? :/

**Comment by adehad** [comment]
> 
> 
> So try compression 1, 2, 3, etc. 9 is the maximum (and the default!). I bet even with 1 or 2 it will be much better.
> I ~suspect~ hope with compression 1 or 2 or 3 you may get a happy balance between speed and size? :/

Compression of 3 seems to be okay for this stimulus, it takes around 50s longer, and the output filesize is now 130MB. Hopefully it will translate well for our longer stimuli as well.
![image](https://user-images.githubusercontent.com/26027314/58792808-c8d9be00-85ec-11e9-93af-45ec3094207c.png)

> 
> Sadly I should have used a real movie format to do this -- a lossless one such as FFv1. But at the time integrating a real movie encoder like ffmpeg was beyond the scope of this project.
> 
I understand! We appreciate your work nonetheless, and I would be happy to look into further development/optimisations with you.

**Comment by cculianu** [comment]
Ok, cool.  Yeah if you come up with changes, do let me know.  For your information: you can probably compile the mex file if you specify all the .cpp files it depends-on on the command-line .. I know it pulls in dependencies from `../../FastMovieFormat.cpp` for the actual fast movie writing... it should work if you pay attention to the error messages (usually it will complain about missing symbols), and add 1 by 1 the .cpp files from the parent dir that are missing.  It also depends on zlib and I think you need to specify those libs too on the command-line...

**Comment by adehad** [comment]
I actually tried running the `.fmv` on StimGL now, using the MATLAB `uint8( )` type casting, the stimulus does not display properly. The entire screen flashes between black and white instead. Allowing the MEX file to handle the conversion works as expected, but at the cost of speed.

**Comment by cculianu** [comment]
Gaah! Really.  That.. shouldn't be the case.  Data is data.  Very strange that it would do that.  Are  you sure `uint8` casting does what it claims to?  Hmm... I honestly forgot my maltlab details..

I'm sorry I don't really have time to work on this now .. I would perhaps fix up the code.  It's strange to me that the data would produce bad movie files in that case.. it makes little sense...

**Comment by cculianu** [comment]
Just double check that the dimensions of the casted `uint8` matrix weren't changed after the cast?

**Comment by adehad** [comment]
It is very strange:
![image](https://user-images.githubusercontent.com/26027314/58888656-a8d6f700-86df-11e9-8f1d-955d8708f6a8.png)

Here we can see our black and white image matrix `bw`. After typecasting to `uint8` (`aaa`) from `double` (`bbb`) we confirm that the two arrays' (i.e. frames) data are equal `isequal(aaa,bbb)` .
The `whos` command confirms the dimensions are equal.
Using the `imagesc()` command (e.g. `imagesc(aaa)`) we were able to confirm that in MATLAB the stimulus is visible in the black and white image, on both the original and typecasted frames.

> I'm sorry I don't really have time to work on this now .. I would perhaps fix up the code. It's strange to me that the data would produce bad movie files in that case.. it makes little sense...

No worries, I am just posting here in case anyone else finds it useful too. If we find a solution I will post that on here too.

**Comment by adehad** [comment]
Just found out it was a big yikes on our side.
The frame we passed was 1 for white and 0 for black, a 'binary image'. But AddFrame expected a 255 to 0 range instead for the `uint8` type. For the `double` and `single` type it actually multiplies by 255:
https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/FastMovieWriterMex/FastMovieWriterMex.cpp#L184-L185
Which is in line with the documentation for the `AddFrame` function, where it expects intensity values:
https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/%40FastMovieWriter/AddFrame.m#L3-L4

So if you are going to make a backwards compatible fix, it might be something like:
```
FastMovieWriterMex('addFrame', g.handle, uint8(255*frame), clevel);
```
instead of:
https://github.com/cculianu/StimulateOpenGL_II/blob/827a64561893a5b5b95521e3f39ff16f765ec1a0/Matlab/%40FastMovieWriter/AddFrame.m#L25

**Comment by cculianu** [comment]
Oh wow.  That should have at least been documented.  GRRR.  Yeah to my developer mind "of course it would be scaled to 255" -- but that fact is not obvious -- esp since the documentation is WAYY misleading.  It also doesn't talk about the performance impact and how uint8 is the fastest.

I agree this needs to be updated.  I'm leaving this issue open and may get to it soon.

So .. I take it now it works, right?

**Comment by adehad** [comment]
> So .. I take it now it works, right?

Yep, it now correctly displays the stimulus after typecasting with `uint8`, sorry I didn't make that clear !

**Comment by cculianu** [comment]
Ok, cool. I'm leaving this issue open.  I may find time soon to go in and better document the behavior and also give this code a little love.. Thanks for posting and do use this issue tracker.  It's probably the best way to get a hold of me and to document things for others.
